### ISSUE TITLE
SAMRAI 3 conversion

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#263]
I have discussed this with @abarret and @boyceg both in person and over email; we should summarize everything here.

I talked to @abarret about this yesterday: since ~~SAMRAI 3.x~~ SAMRAI 3.12 requires C++11 (e.g., `Pointer` is replaced by `std::shared_ptr`) we should merge #261 before switching over.

One major change in the new SAMRAI releases is that most classes were de-templatized for a variety of reasons (IIRC, some classes don't even use the template argument). Here is a brief list of what needs to happen:
- [x] come up with a list of incompatibilities between the two versions (@abarret has this)
- [ ] write a script that can automatically switch most of the constructs (I believe @abarret has this too)
- [ ] examine the necessity of the patches @deleeke wrote: these performance problems have *probably* been fixed over the last decade but we should double check (and send the SAMRAI guys a patch if we need to!)
- [ ] 'press the button' and convert IBAMR to require this newer version (update the library usage and `m4/configure_samrai.m4`)

Since the new versions of SAMRAI are de-templatized, some of the motivation for compiling things in 2D and 3D are now gone. However, since we do (IIRC) compile-time switches between calling one FORTRAN kernel or another I believe it would be best to keep the 2D/3D split for now.

@abarret Please edit this comment if I am missing something.

Edit: It is SAMRAI 3.12 that requires C++11. I believe the present intent is to switch to that version (which is the newest).

**Comment by boyceg** [comment]
> One major change in the new SAMRAI releases is that most classes were de-templatized for a variety of reasons (IIRC, some classes don't even use the template argument). Here is a brief list of what needs to happen:
> 
>  come up with a list of incompatibilities between the two versions (@abarret <https://github.com/abarret> has this)
>  write a script that can automatically switch most of the constructs (I believe @abarret <https://github.com/abarret> has this too)
>  examine the necessity of the patches @deleeke <https://github.com/deleeke> wrote: these performance problems have probably been fixed over the last decade but we should double check (and send the SAMRAI guys a patch if we need to!)
I believe that this performance issue is still present in SAMRAI 3. I reported it when we discovered it earlier in the AY.
>  'press the button' and convert IBAMR to require this newer version (update the library usage and m4/configure_samrai.m4)
You are very optimistic! :-)
> Since the new versions of SAMRAI are de-templatized, some of the motivation for compiling things in 2D and 3D are now gone. However, since we do (IIRC) compile-time switches between calling one FORTRAN kernel or another I believe it would be best to keep the 2D/3D split for now.
> 
I agree. Once we get an initial version working, then we can work on removing this specialization in IBAMR.

**Comment by boyceg** [comment]
>> 
>> One major change in the new SAMRAI releases is that most classes were de-templatized for a variety of reasons (IIRC, some classes don't even use the template argument). Here is a brief list of what needs to happen:
>> 
>>  come up with a list of incompatibilities between the two versions (@abarret has this)
>>  write a script that can automatically switch most of the constructs (I believe @abarret has this too)
>>  examine the necessity of the patches @deleeke wrote: these performance problems have probably been fixed over the last decade but we should double check (and send the SAMRAI guys a patch if we need to!)
> I believe that this performance issue is still present in SAMRAI 3. I reported it when we discovered it earlier in the AY.
>>  'press the button' and convert IBAMR to require this newer version (update the library usage and m4/configure_samrai.m4)
> You are very optimistic! :-)

Although if it is possible to automate the required changes, we should do it both for us and also for users, who also will face pretty substantial code modifications to update to SAMRAI 3.

**Comment by amneetb** [comment]
Yes, please keep us in mind when changing to SAMRAI 3 and C++11. We would really love automation for our existing codes and branches.

**Comment by drwells** [comment]
See https://github.com/IBAMR/IBAMR/tree/abarret/samrai3-migration for the present status.

**Comment by drwells** [comment]
@abarret What is the current state of this?

**Comment by abarret** [comment]
I stalled when I hit a point where we needed `IBTK_MPI`. Should we switch over all instances of `SAMRAI_MPI` to `IBTK_MPI` now, or put that in the same patch as the conversion to SAMRAI 3? I have a script that can do the switch; I could put in a PR today if we wanted to make the switch in 0.6.

**Comment by boyceg** [comment]
That seems like a big change. I think it would be good to get it tested out on some real apps before putting it in a release.

**Comment by abarret** [comment]
> That seems like a big change. 

I would hope that it's only a namespace change. The `IBTK_MPI` code was basically copied from the `SAMRAI_MPI` version. But I can put up a PR when 0.6 is released.

**Comment by boyceg** [comment]
@drwells what do you think?

**Comment by drwells** [comment]
It should definitely not go in 0.6 - now that (almost all!) the patches are in its time to start testing the release.

We can write the patch to do the switch now, but I would prefer to wait on merging it until after we create a release branch (should happen near the end of the week).

**Comment by drwells** [comment]
Time goes by and this issue remains. @abarret I know we hit another wall w.r.t. `Patch::is_periodic()` or whatever it was. Could you post all your notes and whatnot here so that we have everything in one place?

**Comment by drwells** [comment]
@abarret *ping*

**Comment by drwells** [comment]
I want to take another look at this soon. We should try to see if we can resolve the periodicity issue and then try to port minimal chunks of IBTK over.

**Comment by drwells** [comment]
Let me write it out here: we may want to convert this into a checklist.

Updating to SAMRAI main branch:

List of changes:

##### Can be done or detected *mostly* automatically #####

1) SAMRAI is no longer templated on the dimension. Certain classes require the dimension as an input in the constructor if they can not figure out from member variables.
    -- removeTemplateOnDim.pl

2) SAMRAI Arrays are now deprecated and will be removed in the future. You should use the std::vector instead as it supplies the same functionality.
    -- changeArray.pl

3) SAMRAI MPI calls are no longer static functions. To make an MPI call using the SAMRAI wrappers now requires an SAMRAI_MPI object.
    -- IBTK_MPI class exists. Already changed in IBAMR.
    -- update_mpi_calls.pl already in repository

4) Box and Patch Iterators have been changed to use the now common begin() and end() notation.
    -- updateIterators.pl searches for for() loops
    -- Fragile script, requires for() loops to be written as in IBAMR.

5) The SAMRAI Pointer class has been eliminated. Use std::shared_ptr instead.
    -- removePointer.pl replaces Pointer with shared_ptr and new with make_shared
    -- Doesn't detect 'new' usage that doesn't go with Pointer

6) Since the Pointer class was removed, implicit casting between PatchData and Cell|Node|Edge|Face|SideData types and PatchGeometry and CartesianPatchGeometry is no longer allowed.
    -- use_static_pointer_cast.pl searches for implicit conversion.
    -- Possible other implicit casting that we can detect?
        -- Variable types

7) The VariableDatabase can no longer register variable and context pairs with no ghost cell width since the IntVector constructor now requires a Dimension argument.
    -- changeVariableDatabase.pl searches for registerVariableAndContext() with empty IntVectors()

8) The base files Base(PatchHierarchy|PatchLevel|GriddingAlgorithm) no longer exist.
    -- baseFiles.pl replaces Base* object declarations with * declarations.

9) LoadBalancer has been changed to ChopAndPackLoadBalancer.
    -- loadBalancer.pl replaces LoadBalancer with ChopAndPackLoadBalancer.

10) SAMRAI include file locations have changed to prevent collisions.
    -- convertIncludePaths.pl replaces SAMRAI includes with new locations.
    -- Does not detect possible collisions.
    -- SAMRAI m4 files are now located at SAMRAI/pdat/fortran

11) Certain GriddingAlgorithm characteristics pertaining to levels have been moved to the PatchHierarchy.
    -- e.g. getMaxNumberOfLevels() now are in PatchHierarchy.
    -- make_list_of_change_to_patchHierarchy.pl creates a text file of possible defunct GriddingAlgorithm calls.

##### No script to do automatically #####

12) The BoxList and BoxTree classes have been merged into the single class BoxContainer.
    -- Can't be done automatically. but make_list_of_box.pl creates a list of occurences.

13) SAMRAI startup and shutdown interfaces have changed.
    -- IBTKInit() has been created and used in examples.

14) The interface for RefineOperator::refine() has changed.
    -- "fine_box" is now a BoxOverlap.

15) Input databases for GriddingAlgorithm, PatchHierarchy, and BergerRigoutsos have changed.
    -- Information about the patch levels in the input database for GriddingAlgorithm have moved to PatchHierarchy.
    -- Information about the box generation strategy such as efficiency tolerance have moved to BergerRigoutsos input database.
    -- Note that now PatchHierarchy and BergerRigoutsos now take input databases as an input in the constructor.

16) The REFINE_BOXES option in the StandardTagAndInitialize database has changed structure.

17) Autoconf-based build system has been removed.
    -- Now uses CMake supplemented with the BLT macro library.

##### Items that weren't mentioned in release notes or I didn't notice #####

18) Underlying data storage for PatchData now uses std::vector
    -- PatchData can no longer be used with bool as the data type

19) Serializable::writeToDatabase() is now a const function.
    -- LData::writeToDatabase() calls non-const PETSc functions

20) PatchGeometry::getTouchesPeriodicBoundary() no longer exists.
    -- In particular LEInteractor relies on this to determine what kind of boundary it's on.
    -- LEInteractor needs this functionality for when IB points are outside the patch, but their interaction regions intersect with the patch.

21) (Coarsen|Refine)Operator is now in SAMRAI::hier instead of SAMRAI::xfer

22) (Coarsen|Refine)Operator does not have findCoarsenOperator().

23) (Coarsen|Refine)Operator getOperatorName() is not virtual.
    -- The operator name is now set by the constructor and is private.

24) (Coarsen|Refine)Operator and RefinePatchStrategy getStencilWidth() now takes a Dimension argument.

25) Registering coarsen/refine operators changed.

**Comment by drwells** [comment]
@abarret could you add a single script (either in shell or perl) that runs all the other scripts?

**Comment by abarret** [comment]
> @abarret could you add a single script (either in shell or perl) that runs all the other scripts?

Yes. There's already one in the branch.

**Comment by drwells** [comment]
Oh - which one is it?

**Comment by abarret** [comment]
update_to_samrai_3 or something very similar. I probably should have added an extension. :)

**Comment by drwells** [comment]
@abarret Could you move that branch into this repo? I want to start working on it and it will be easier to track this way.

**Comment by abarret** [comment]
The branch is now located at https://github.com/IBAMR/IBAMR/tree/samrai_main

**Comment by abarret** [comment]
@drwells, how do you want to manage changes to the branch? Should I push changes to my own repository, or this repository?

**Comment by drwells** [comment]
I'd prefer if we put up PRs here so I can see the progress.

**Comment by drwells** [comment]
The other advantage of PRs is that we can concurrently write patches without running into problems.

**Comment by abarret** [comment]
Do you have a suggested commit strategy? My current strategy is to commit whenever I have a subdirectory that builds.

**Comment by drwells** [comment]
I'd rather not check in changes to the library itself yet - we should just continue updating the scripts.

**Comment by drwells** [comment]
In particular - if we maintain a branch where the scripts have been executed then that branch will always have difficult to understand conflicts with `master`. We should never check in those changes - instead we should check in everything else (like changes to the build system).
